test/model/message_test.dart

```diff
@@ -524,24 +529,148 @@ void main() {
+        ) {
+          if (displayRecipient == null) throw ArgumentError();
+          json['display_recipient'] = displayRecipient;
+        }
+
+        return Message.fromJson(json);
+      }
+
+      /// Checks if the single message in [MessageStore.messages]
+      /// is identical to [message].
+      void checkStoredMessageIdenticalTo(Message message) {
+        check(store.messages)
+          .deepEquals({message.id: conditionIdentical(message)});
+      }
+
+      test('DM', () async {
+        await prepare();
+        final message = eg.dmMessage(id: 1, from: eg.otherUser, to: [eg.selfUser]);
+
+        store.reconcileMessages([message]);
+        checkStoredMessageIdenticalTo(message);
+        store.reconcileMessages([copyStoredMessage()]);
+        // Not clobbering, because the first call didn't mark stale.
+        checkStoredMessageIdenticalTo(message);
+      });
+
+      group('channel message; chooses correctly whether to clobber the stored version', () {
+        // Exercise the ways we move the message in and out of the "maybe stale"
+        // state. These include reconcileMessage itself, so sometimes we test
+        // repeated calls to that with nothing else happening in between.
+
+        test('various conditions', () async {
+          final channel = eg.stream();
+          await prepare(stream: channel, isChannelSubscribed: true);
+          final message = eg.streamMessage(id: 1, stream: channel);
+
+          final otherChannel = eg.stream();
+          await store.addStream(otherChannel);
+
+          store.reconcileMessages([message]);
+          checkStoredMessageIdenticalTo(message);
+          store.reconcileMessages([copyStoredMessage()]);
+          // Not clobbering, because the first call didn't mark stale,
+          // because the message was in a subscribed channel.
+          checkStoredMessageIdenticalTo(message);
+
+          await store.removeSubscription(channel.streamId);
+          messageCopy = copyStoredMessage();
+          store.reconcileMessages([messageCopy]);
+          // Clobbering because the unsubscribe event marked the message stale.
+          checkStoredMessageIdenticalTo(messageCopy);
+          messageCopy = copyStoredMessage();
+          store.reconcileMessages([messageCopy]);
+          // (Check that reconcileMessage itself didn't unmark as stale.)
+          checkStoredMessageIdenticalTo(messageCopy);
+
+          await store.addSubscription(eg.subscription(channel));
+          messageCopy = copyStoredMessage();
+          store.reconcileMessages([messageCopy]);
+          // The channel became subscribed,
+          // but the message's data hasn't been refreshed, so clobber…
+          checkStoredMessageIdenticalTo(messageCopy);
+
+          store.reconcileMessages([copyStoredMessage()]);
+          // …Now it's been refreshed, by reconcileMessages, so don't clobber.
+          checkStoredMessageIdenticalTo(messageCopy);
```

Comment:
This test case is a bit long and fiddly to read. I think it could be simplified with a pair of helpers along these lines:
```suggestion
          // The channel became subscribed,
          // but the message's data hasn't been refreshed, so clobber…
          checkClobber();

          // …Now it's been refreshed, by reconcileMessages, so don't clobber.
          checkNoClobber();
```

which should make it easier to scan and see all the different scenarios it's going through.